File: pkg/ai/random/random.go
package random

import (
	"math/rand"
	"skate-area-battle-game/pkg/game"
)

// RandomAI はランダムに行動を選ぶ実装
type RandomAI struct{}

// New は RandomAI を生成する
func New() game.AI { return &RandomAI{} }

func (r *RandomAI) SelectBoard(states []*game.GameState) int {
	return rand.Intn(len(states))
}
func (r *RandomAI) SelectTurn(states []*game.GameState) int {
	return rand.Intn(2)
}
func (r *RandomAI) Evaluate(state *game.GameState, player int) float64 {
	// 評価: 自分の塗りマス数 - 相手の塗りマス数
	var me, opp int
	for y := 0; y < state.N; y++ {
		for x := 0; x < state.N; x++ {
			if state.Owner[y][x] == player {
				me += state.Value[y][x]
			} else if state.Owner[y][x] == 1-player {
				opp += state.Value[y][x]
			}
		}
	}
	return float64(me - opp)
}

File: pkg/ai/sneuaiolake/trainer/game/src/ai/random.rs
use anyhow::Result;
use rand::{Rng, SeedableRng};
use rand_xorshift::XorShiftRng;

use super::AI;
use crate::game::GameState;

#[derive(Debug)]
pub struct RandomAI {
    rng: XorShiftRng,
}

impl RandomAI {
    pub fn new() -> Self {
        RandomAI {
            rng: XorShiftRng::from_os_rng(),
        }
    }
}

impl Default for RandomAI {
    fn default() -> Self {
        Self::new()
    }
}

impl AI for RandomAI {
    fn name(&self) -> &str {
        "random"
    }

    fn select_board(&mut self, states: &[GameState]) -> Result<usize> {
        Ok(self.rng.random_range(0..states.len()))
    }

    fn select_turn(&mut self, _states: &[GameState]) -> Result<usize> {
        Ok(self.rng.random_range(0..2))
    }

    fn evaluate(&mut self, _state: &GameState, _player: usize) -> Result<f32> {
        Ok(self.rng.random_range(-1.0..1.0))
    }
}
